### PR TITLE
946 generation queue

### PR DESCRIPTION
**Pull request by rpoletaev**

### CODE REVIEW

File: packages/daemons/block_generator.go [hunk 1]
```diff
@@ -61,13 +62,13 @@ func BlockGenerator(ctx context.Context, d *daemon) error {
 		return err
 	}
 
-	blockTimeCalculator, err := utils.BuildBlockTimeCalculator(nil)
-	if err != nil {
-		d.logger.WithFields(log.Fields{"type": consts.BlockError, "error": err}).Error("building block time calculator")
-		return err
-	}
+	// blockTimeCalculator, err := utils.BuildBlockTimeCalculator(nil)
```
Comment (potehinre) [inline packages/daemons/block_generator.go]: commented code prohibited

File: packages/daemons/block_generator.go [hunk 2]
```diff
@@ -61,13 +62,13 @@ func BlockGenerator(ctx context.Context, d *daemon) error {
...
+	// if err != nil {
+	// 	d.logger.WithFields(log.Fields{"type": consts.BlockError, "error": err}).Error("building block time calculator")
+	// 	return err
+	// }
 
-	timeToGenerate, err := blockTimeCalculator.SetClock(&utils.ClockWrapper{}).TimeToGenerate(nodePosition)
+	timeToGenerate, err := protocols.TimeToGenerate(time.Now(), int(nodePosition))
```
Comment (potehinre) [inline packages/daemons/block_generator.go]: i think that should be like protocols.BlockGenerationTime.TimeToGenerate

File: packages/daemons/block_generator.go [hunk 3]
```diff
@@ -119,7 +120,7 @@ func BlockGenerator(ctx context.Context, d *daemon) error {
 		Version:      consts.BLOCK_VERSION,
 	}
 
-	timeToGenerate, err = blockTimeCalculator.SetClock(&utils.ClockWrapper{}).TimeToGenerate(nodePosition)
+	timeToGenerate, err = protocols.TimeToGenerate(time.Now(), int(nodePosition))
```
Comment (potehinre) [inline packages/daemons/block_generator.go]: same stuff

File: packages/protocols/generation-queue.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	"time"
+
+	"github.com/GenesisKernel/go-genesis/packages/model"
+
+	"github.com/GenesisKernel/go-genesis/packages/conf/syspar"
+	log "github.com/sirupsen/logrus"
+)
+
+// QueueChecker allow check queue to generate current block
+type QueueChecker interface {
```
Comment (potehinre) [inline packages/protocols/generation-queue.go]: And again, what queue does it actually check?
All functions of that interface is about block generation time.
Maybe better name it something like BlockTimeChecker

File: packages/protocols/generation-queue.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	TimeToGenerate(position int64) (bool, error)
+	NextTime(position int64, t time.Time) (time.Time, error)
+	BlockForTimeExists(t time.Time, nodePosition int) (bool, error)
+	RangesByTime(t time.Time) (start, end time.Time)
```
Comment (potehinre) [inline packages/protocols/generation-queue.go]: RangeSByTime, but i see only one range

File: packages/protocols/generation-queue.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+type BlockTimeCounter struct {
+	start       time.Time
+	duration    time.Duration
+	numberNodes int
+}
+
+// Queue returns serial queue number for time
+func (btc *BlockTimeCounter) Queue(t time.Time) (int, error) {
```
Comment (potehinre) [inline packages/protocols/generation-queue.go]: BlockID(t time.Time) ?

File: packages/protocols/generation-queue.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+	b := &model.Block{}
+	blocks, err := b.GetNodeBlocksAtTime(startInterval, endInterval, int64(nodePosition))
+	if err != nil {
+		return false, err
+	}
+
+	if len(blocks) != 0 {
+		log.WithFields(log.Fields{"type": "block_time_counter", "error": DuplicateBlockError, "start": startInterval, "end": endInterval}).Error("")
+		return false, DuplicateBlockError
```
Comment (potehinre) [inline packages/protocols/generation-queue.go]: well, it exists for that time, why false?

File: packages/protocols/protocols.go [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import "time"
+
+// TimeToGenerate allow check block generation time for nodePosition
+func TimeToGenerate(at time.Time, nodePosition int) (bool, error) {
+	tbc := NewBlockTimeCounter()
+	return tbc.TimeToGenerate(at, nodePosition)
+}
+
+func BlockForTimeExists(t time.Time, nodePosition int) (bool, error) {
+	btc := NewBlockTimeCounter()
```
Comment (potehinre) [inline packages/protocols/protocols.go]: i think it should be used outside that function

File: packages/protocols/protocols.go [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package protocols
+
+import "time"
+
+// TimeToGenerate allow check block generation time for nodePosition
+func TimeToGenerate(at time.Time, nodePosition int) (bool, error) {
+	tbc := NewBlockTimeCounter()
```
Comment (potehinre) [inline packages/protocols/protocols.go]: same stuff

File: packages/protocols/generation-queue.go (hunk 7 above)
Comment (potehinre) [reply packages/protocols/generation-queue.go]: Did you mean BlockForTimeNotExists?

File: packages/protocols/generation-queue.go (hunk 5 above)
Comment (potehinre) [reply packages/protocols/generation-queue.go]: And second, it used only internally in that object, why do we need it in interface?

**Review by potehinre — CHANGES_REQUESTED** [review]

File: packages/protocols/generation-queue.go (hunk 6 above)
Comment (rpoletaev) [reply packages/protocols/generation-queue.go]: this func returns serial number of generation time interval, that not equal blockID

**Review by potehinre — APPROVED** [review]

**Pull request merged**